src/pint/toa.py

```diff
@@ -1417,6 +1427,49 @@ def compute_posvels(self, ephem=None, planets=False):
         log.debug("Adding columns " + " ".join([cc.name for cc in cols_to_add]))
         self.table.add_columns(cols_to_add)
 
+    def add_vel_ecl(self):
+        """Compute and add a column to self.table with velocities in ecliptic coordinates.
+
+        Called in barycentric_radio_freq() in AstrometryEcliptic (astrometry.py) 
+        if ssb_obs_vel_ecl column does not already exist.
+        If compute_posvels() called again for a TOAs object (aka TOAs modified), 
+        deletes this column so that this function will be called again and 
+        velocities will be calculated with updated ssb_obs.
+        """
+        # Remove any existing columns
+        col_to_remove = "ssb_obs_vel_ecl"
+        if col_to_remove in self.table.colnames:
+            self.table.remove_column(col_to_remove)
+
+        ssb_obs_vel_ecl = table.Column(
+            name="ssb_obs_vel_ecl",
+            data=np.zeros((self.ntoas, 3), dtype=np.float64),
+            unit=u.km / u.s,
+            meta={"origin": "SSB", "obj": "OBS"},
+        )
+
+        # Now step through in observatory groups
+        for ii, key in enumerate(self.table.groups.keys):
+            loind, hiind = self.table.groups.indices[ii : ii + 2]
+            ssb_obs = self.ssb_obs
+            # convert ssb_obs pos and vel to ecliptic coordinates
+            coord = ICRS(
+                x=ssb_obs.pos[0],
+                y=ssb_obs.pos[1],
+                z=ssb_obs.pos[2],
+                v_x=ssb_obs.vel[0],
+                v_y=ssb_obs.vel[1],
+                v_z=ssb_obs.vel[2],
+                representation_type=CartesianRepresentation,
+                differential_type=CartesianDifferential,
+            )
+            coord = coord.transform_to(PulsarEcliptic)
+            # get velocity vector from coordinate frame
+            ssb_obs_vel_ecl[loind:hiind, :] = coord.velocity.d_xyz.T.to(u.km / u.s)
+        col = ssb_obs_vel_ecl
+        log.debug("Adding columns " + " ".join(col.name))
+        self.table.add_column(col)
```

Comment:
Why should `add_vel_ecl()` be in `observatory`?  It is right next to `compute_posvels()` here in toa.py.  Do you think both should be moved?  The ecliptic velocity calculation could have been just added to compute_posvels() but then it would have been computed unnecessarily for all par files that were in Equatorial coordinates. It is separate just so it can be computed only in situations where it is actually needed (lazy initialization).

Reply:
There are two reasons  1. We are recomputing the obs ICRS coordinate here in this function. however, the observatory module already did it. 2. Those locations depend on the par file parameter ECL. If we want to apply the same TOA to a new parfile with a different ECL.  Those columns have to be recomputed. I actually like your idea of putting add_vel_ecl() in the TOA class. I suggest we put the skycoor ICRS object in the table, so we don't need to recompute the ICRS sky coordinate. We can also call add_vel_ecl() with argument ECL.

Furthermore, (I am not sure if this is a good idea) we can only store the astropy coordinate object and make the obs_ssb_posvel as properties. But it is not in the scope of this PR.

Reply:
Why would we put the ICRS sky coordinate in the TOAs table?  That is something that is usually fit for and so constantly changes.  Or am I missing something?

Reply:
@scottransom and @paulray I meant the earth location in ICRS Coordinate.

Reply:
I don't think you want to do that.  Some observatories are not linked to the earth (like NICER!)

Reply:
Good point. What I am trying to do is to make the ICRS object, line 1456, once. So that in the future, we don't have to make it again if more transformations are needed.

Reply:
I don't think we need to implement caching here. This routine is called only once to add the column to the TOAs table.  It is not going to be called repeatedly, so the risk of invalid caches is greater than any benefit, in my opinion.

Reply:
No, we should not cache anything here. I was thinking of returning an ICRS object from compute_posvel(). I think it will for another PR's discussion.

Reply:
Oh, I see. Yeah, that might be worth considering. When PosVels were defined astropy surely didn't support full positions with velocities as it does now.  Agree that is best considered for a separate PR.